examples/4_secret_storage.py

```diff
@@ -0,0 +1,377 @@
+        team: str,
+        namespace: str,
+        entryKey: str,
+        entryValue: str,
+        revision: Union[int, None] = None,
+    ) -> keybase1.KVPutResult:
+        h = await self.hmac_key(secret, entryKey)
+        res = await self.kvstore.put(team, namespace, h, entryValue, revision)
+        res.entry_key = entryKey
+        return res
+
+    async def delete(
+        self,
+        secret: bytes,
+        team: str,
+        namespace: str,
+        entryKey: str,
+        revision: Union[int, None] = None,
+    ) -> keybase1.KVDeleteEntryResult:
+        h = await self.hmac_key(secret, entryKey)
+        res = await self.kvstore.delete(team, namespace, h, revision)
+        res.entry_key = h
+        return res
+
+    async def get(
+        self, secret: bytes, team: str, namespace: str, entryKey: str
+    ) -> keybase1.KVGetResult:
+        h = await self.hmac_key(secret, entryKey)
+        res = await self.kvstore.get(team, namespace, h)
+        res.entry_key = h
+        return res
+
+    async def list_entrykeys(
+        self, secret: bytes, team: str, namespace: str
+    ) -> keybase1.KVListEntryResult:
+        res = await self.kvstore.list_entrykeys(team, namespace)
+        if res.entry_keys:
+            for e in res.entry_keys:
+                if not e.entry_key.startswith("_"):
+                    get_res = await self.kvstore.get(team, namespace, e.entry_key)
+                    e.entry_key = json.loads(get_res.entry_value)[
+                        self.KEY_KEY
+                    ]  # modify
+        return res
+
+
+class RentalHandler:
+    """
+    RentalHandler handles commands sent via chat to use the team key-value store.
+
+    RentalHandler listens to chat messages of the form:
+    `!rental {reserve|unreserve} <tool> <YYYY-MM-DD>`
+    `!rental {lookup|add|remove} <tool>`
+    `!rental list` // lists all tools
+
+    It expects properly formatted commands.
+
+    RentalHandler uses a cache to keep track of the most recently fetched value and
+    revision for each key. To handle concurrent updates, it attempts to update
+    with the most recently fetched revision + 1; if it fails, it does a "get"
+    and updates the cache, and asks the user to retry if they still want to
+    do their update.
+
+    RentalHandler also maintains a cache of the per-team per-namespace hmac secrets
+    it has previously fetched (stored with the special entryKey "_secret"),
+    which it uses to HMAC entryKeys. RentalHandler uses the SecretKeyKVStoreClient
+    to store entryKeys' HMACs instead of the plaintext entryKeys.
+   """
+
+    MSG_PREFIX = "!rental"
+    NAMESPACE = "rental"
+    SECRET_NUM_BYTES = 32
+    SECRET_KEY = "_secret"
+
+    def __init__(self):
+        # self.cache = {tool: {"revision": int, "info": {} or None}}
+        self.cache: Dict[Any, Any] = {}
+        self.secrets: Dict[str, Dict[str, bytes]] = {}  # {team: {namespace: secret}}
+
+    async def load_secret(self, bot, team, namespace):
```

Comment:
i'm not exactly sure where to leave this comment, so i'm just going to do it here!
the way you've separated the responsibilities between RentalHandler and SecretKeyKVStoreClient means that RentalHandler is doing a lot of the _secret_-y work. so like, if we were to create another handler that also wanted to have secret entryKeys, it would have to reimplement this (load_secret) method. 

i'd suggest considering a more object-y approach to this. the secret client doesn't need to be stateless. in fact, i'd argue it should manage as much of this process as possible so the rentalhandler doesn't need to know about secrets at all. it just gets and puts through the secret-kv-client. i'd actually say ideally the secretClient should have the exact same api as the normal kvstoreclient (which you've basically done) and the rental handler should be able to call it exactly the same way. yeah - so the secretkvclient should make it as easy as possible to take a working, normal handler and turn it into a secret one by dropping in this new client: couple of steps like maybe initializing it with some randomness (maybe not necessary), assigning it to something somewhere, and boom - off to the races.

Reply:
i moved out the secrets logic in RentalHandler and made a a custom Bot class that has a secret_kvstore() property, and stores secrets state

Reply:
Looks great to me :)